src/ir/auth_logic/lowering_ast_datalog.h

```diff
@@ -0,0 +1,128 @@
+    int fresh_var_count_;
+
+    // This returns a fresh variable name. Fresh variables are introduced in a
+    // few places during the translation.
+    std::string FreshVar() {
+      return "x__" + std::to_string(fresh_var_count_++);
+    }
+
+    // This function makes a copy of a predicate that prepends the name with
+    // another string and prepends the list of arguments with additional 
+    // arguments. This is used in a few places in the translation, for
+    // example, to translate "X says blah(args)" into "says_blah(X, args)".
+    Predicate PushOntoPred(std::string modifier,
+        std::vector<std::string> new_args, const Predicate& predicate) {
+      std::string new_name = modifier + predicate.name();
+      new_args.insert(new_args.end(), predicate.args().begin(),
+          predicate.args().end());
+      Sign sign_copy = predicate.sign();
+      return Predicate(new_name, new_args, sign_copy);
+    }
+
+    Predicate PushPrin(std::string modifier,
+        const Principal& principal, const Predicate& predicate) {
+      return PushOntoPred(modifier, {principal.name()}, predicate);
+    }
+
+    Predicate AttributeToDLIR(const Attribute& attribute) {
+      // If attribute is `X pred(args...)` the following predicate is
+      // `pred(X, args...)`
+      return PushPrin(std::string(""), attribute.principal(),
+          attribute.predicate());
+    }
+
+    // This produces an extra rule that needs to be generated during the 
+    // translation of "speaker says attribute" into DLIR
+    DLIRCondAssertion SpokenAttributeToDLIR(const Principal& speaker,
+        const Attribute& attribute) {
+      Predicate main_predicate = AttributeToDLIR(attribute);
+
+      // Attributes interact with "canActAs" because if "Y canActAs X"
+      // then Y also picks up X's attributes. We need to generate
+      // an additional rule to implement this behavior. If the attribute
+      // under translation is `X PredX`, the additional rule is:
+      // `speaker says Y PredX :-
+      //    speaker says Y canActAs X, speaker says X PredX`
+      // (Where Y is a fresh variable)
+      Principal prin_y = Principal(FreshVar());
+      
+      // This is `speaker says Y PredX`
+      Predicate generated_lhs = PushPrin(std::string("says_"),
+          prin_y, main_predicate);
+
+      Predicate y_can_act_as_x = Predicate(std::string("canActAs"),
+          {prin_y.name(), attribute.principal().name()},
+          Sign::kPositive);
+
+      Predicate speaker_says_y_can_act_as_x = PushPrin(
+          std::string("says_"), speaker, y_can_act_as_x);
+
+      // This is `speaker says X PredX`
+      Predicate speaker_says_x_pred = PushPrin(std::string("says_"),
+          speaker, main_predicate);
+
+      // This is the full generated rule:
+      // `speaker says Y PredX :-
+      //    speaker says Y canActAs X, speaker says X PredX`
+      return DLIRCondAssertion(generated_lhs,
+          {speaker_says_y_can_act_as_x, speaker_says_x_pred});
```

Comment:
As the final uses of `generated_lhs`, `speaker_says_y_can_act_as_x`, and `speaker_says_x_pred`, these should use `std::move`.

Reply:
fixed